### PR TITLE
IGNITE-20685 Implement ability to trigger transaction recovery

### PR DESCRIPTION
**Pull request by vldpyatkov**
https://issues.apache.org/jira/browse/IGNITE-20685

### CODE REVIEW

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                        .build());
+            }
+
+            return CompletableFuture.completedFuture(false);
+        };
+    }
+
+    @Override
+    public void start() {
+        lockManager.listen(LockEvent.CONFLICT_FOUND, lockConflictListener);
```
Comment (denis-chudov) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java:103]: I thnk a method reference could work fine here, no need to add the field `lockConflictListener`

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            LockManager lockManager,
+            TxManager txManager,
+            ClusterService clusterService
+    ) {
+        this.lockManager = lockManager;
+        this.txManager = txManager;
+        this.clusterService = clusterService;
+
+        lockConflictListener = (parameters, exception) -> {
+            TxStateMeta txSate = txManager.stateMeta(parameters.getLockHolderTx());
```
Comment (denis-chudov) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java:80]:
```suggestion
            TxStateMeta txState = txManager.stateMeta(parameters.getLockHolderTx());
```

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+            if (txSate == null || clusterService.topologyService().getById(txSate.txCoordinatorId()) == null) {
+                LOG.info("Conflict was found, and the coordinator of the transaction that holds a lock is not available "
+                                + "[lockHolderTx={}, lockAcquirerTx={}].",
```
Comment (denis-chudov) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java:84]: pls add coordinator id

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                        parameters.getLockHolderTx(),
+                        parameters.getLockAcquirerTx()
+                );
+
+                //TODO:IGNITE-20735 Obviously, the commit partition node does not have to be at the local node.
+                ClusterNode cmpPartNode = clusterService.topologyService().localMember();
```
Comment (denis-chudov) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java:90]:
```suggestion
                ClusterNode commitPartNode = clusterService.topologyService().localMember();
```

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            if (msg instanceof TxRecoveryMessage) {
+                var txRecoveryMsg = (TxRecoveryMessage) msg;
+
+                TxStateMeta txState = txManager.stateMeta(txRecoveryMsg.txId());
+
+                if (txState == null || !TxState.isFinalState(txState.txState())) {
+                    assert txState == null || clusterService.topologyService().getById(txState.txCoordinatorId()) == null :
+                            "Transaction coordinator is alive [tx=" + txRecoveryMsg.txId() + ", crd=" + txState.txCoordinatorId() + ']';
+
+                    LOG.warn("Transaction have to be aborted [tx={}].", txRecoveryMsg.txId());
```
Comment (denis-chudov) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java:115]:
```suggestion
                    LOG.warn("Transaction has to be aborted [tx={}].", txRecoveryMsg.txId());
```

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            TxStateMeta txSate = txManager.stateMeta(parameters.getLockHolderTx());
+
+            if (txSate == null || clusterService.topologyService().getById(txSate.txCoordinatorId()) == null) {
+                LOG.info("Conflict was found, and the coordinator of the transaction that holds a lock is not available "
+                                + "[lockHolderTx={}, lockAcquirerTx={}].",
+                        parameters.getLockHolderTx(),
+                        parameters.getLockAcquirerTx()
+                );
+
+                //TODO:IGNITE-20735 Obviously, the commit partition node does not have to be at the local node.
```
Comment (denis-chudov) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxRecoveryManager.java:89]: The ticket says:

`Transactions X that detects (detection logic will be covered in a separate ticket) that coordinator is dead awaits commitPartition primary replica and sends initiateRecoveryReplicaRequest to it in a fully asynchronous manner`

Why do you just pick a local node? IMO the part of sending the initiate recovery message should be fully covered by this PR. The mentioned ticket IGNITE-20735 hasn't anything in description about this part of work.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/table/ItTransactionConflictTest.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        );
+
+        assertThat(primaryReplicaFut, willCompleteSuccessfully());
+
+        String leaseholder = primaryReplicaFut.join().getLeaseholder();
+
+        IgniteImpl commitPartNode = IntStream.range(0, initialNodes()).mapToObj(this::node).filter(n -> leaseholder.equals(n.name()))
+                .findFirst().get();
+
+        log.info("Transaction coordinator is determined [node={}].", commitPartNode.name());
```
Comment (vldpyatkov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/table/ItTransactionConflictTest.java:87]: Mistake here, it is a commit partiton node.

**Pull request closed without merging**